### PR TITLE
Reorganize pass-pipeline for Polly

### PR DESCRIPTION
**Pull request by MatthiasJReisinger**
We now move Polly after LoopRotate since at this point the LLVM IR is more amenable to Polly. For example, a simple gemm kernel like

```
@polly function gemm!(A,B,C)
    m,n = size(A)
    n,o = size(B)
    @inbounds for i=1:m, j=1:o, k=1:n
        C[i,j] += A[i,k] * B[k,j]
    end
end
```

could not be optimized before this change. However, this new position after LoopRotate also makes it necessary to rerun InstCombine before Polly to get rid of hindering phi instructions that are introduced by LCSSA. Additionally, we add Polly's CodePreparation and CodegenCleanUp passes. Parts of these changes have already been discussed in https://groups.google.com/forum/#!topic/polly-dev/P2RZUlKRo0I. The reason why they were held back was to see if further adaptions to the pass-pipeline will become apparent, but for now this configuration seems to be convenient.

@tobig @vtjnash @timholy

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -54,7 +54,7 @@ FLAGS += -I$(shell $(LLVM_CONFIG_HOST) --includedir)
 LLVM_LIBS := all
 ifeq ($(USE_POLLY),1)
 LLVMLINK += -lPolly -lPollyISL
-FLAGS += -I$(shell $(LLVM_CONFIG_HOST) --src-root)/tools/polly/include
+FLAGS += -I$(shell $(LLVM_CONFIG_HOST) --src-root)/tools/polly/include -I$(shell $(LLVM_CONFIG_HOST) --obj-root)/tools/polly/include
```
Comment (tobig) [inline src/Makefile]: Why are makefile changes necessary to reorganize the pass pipeline?

**Comment by ViralBShah** [comment]
Let's merge this after branching for 0.5.

File: src/jitlayers.cpp [hunk 2]
```diff
@@ -165,12 +167,17 @@ void addOptimizationPasses(PassManager *PM)
...
+    PM->add(createLoopIdiomPass()); //// ****
+    PM->add(createLoopRotatePass());           // Rotate loops.
 #ifdef USE_POLLY
+    // LCSSA (which has already run at this point due to the dependencies of the
+    // above passes) introduces redundant phis that hinder Polly. Therefore we
+    // run InstCombine here to remove them.
+    PM->add(createInstructionCombiningPass());
+    PM->add(polly::createCodePreparationPass());
     polly::registerPollyPasses(*PM);
+    PM->add(polly::createCodegenCleanupPass());
```
Comment (tobig) [inline src/jitlayers.cpp]: Codegen cleanup and codegen prepare passes are partially independent changes. I would suggest to add them in separate commits (may be the still pull-request though).

File: src/Makefile (hunk 1 above)
Comment (MatthiasJReisinger) [inline src/Makefile]: To use the CodePreparationPass in jitlayers.cpp it was necessary to include polly/LinkAllPasses.h which in turn includes polly/Config/config.h which is placed in $(shell $(LLVM_CONFIG_HOST) --obj-root)/tools/polly/include.

Comment (tobig) [inline src/Makefile]: On 07/29/2016 01:10 PM, Matthias J. Reisinger wrote:
> In src/Makefile
> <https://github.com/JuliaLang/julia/pull/17696#discussion_r72776288>:
> 
>> @@ -54,7 +54,7 @@ FLAGS += -I$(shell $(LLVM_CONFIG_HOST) --includedir)
>>  LLVM_LIBS := all
>>  ifeq ($(USE_POLLY),1)
>>  LLVMLINK += -lPolly -lPollyISL
>> -FLAGS += -I$(shell $(LLVM_CONFIG_HOST) --src-root)/tools/polly/include
>> +FLAGS += -I$(shell $(LLVM_CONFIG_HOST) --src-root)/tools/polly/include -I$(shell $(LLVM_CONFIG_HOST) --obj-root)/tools/polly/include
> 
> To use the CodePreparationPass in jitlayers.cpp it was necessary to
> include polly/LinkAllPasses.h which in turn includes
> polly/Config/config.h which is placed in $(shell $(LLVM_CONFIG_HOST)
> --obj-root)/tools/polly/include.

OK.

Tobias

**Comment by MatthiasJReisinger** [comment]
@tobig thank you for the feedback! I will address this and update the patch. @ViralBShah I agree, it's not necessary to merge this before 0.5.

**Comment by vtjnash** [comment]
lgtm. Since this is all behind preprocessor macros anyways, I see no particular reason it matters if this merges before the v0.5 branch point or after.

**Pull request merged**